contrib/mesos/pkg/executor/service/podsource.go

```diff
@@ -32,97 +32,89 @@ const (
-	out            chan<- interface{}         // out is the sink for merged pod snapshots
-	mirrorPods     chan []*api.Pod            // mirrorPods communicates snapshots of the current set of mirror pods
-	execUpdates    <-chan kubetypes.PodUpdate // execUpdates receives snapshots of the current set of mesos pods
-}
+type (
+	podName struct {
+		namespace, name string
+	}
+
+	sourceMesos struct {
+		stop          <-chan struct{}
+		out           chan<- interface{} // never close this because pkg/util/config.mux doesn't handle that very well
+		registry      executor.Registry
+		priorPodNames map[podName]string // map podName to taskID
+	}
+)
 
-// newSourceMesos creates a pod config source that merges pod updates from
-// mesos (via execUpdates), and mirror pod updates from the apiserver (via
-// podWatch) writing the merged update stream to the out chan. It is expected
-// that execUpdates will only ever contain SET operations. The source takes
-// ownership of the sourceFinished chan, closing it when the source terminates.
-// Source termination happens when the execUpdates chan is closed and fully
-// drained of updates.
 func newSourceMesos(
-	sourceFinished chan struct{},
-	execUpdates <-chan kubetypes.PodUpdate,
+	stop <-chan struct{},
 	out chan<- interface{},
 	podWatch *cache.ListWatch,
+	registry executor.Registry,
 ) {
 	source := &sourceMesos{
-		sourceFinished: sourceFinished,
-		mirrorPods:     make(chan []*api.Pod),
-		execUpdates:    execUpdates,
-		out:            out,
+		stop:          stop,
+		out:           out,
+		registry:      registry,
+		priorPodNames: make(map[podName]string),
 	}
-	// reflect changes from the watch into a chan, filtered to include only mirror pods (have an ConfigMirrorAnnotationKey attr)
-	cache.NewReflector(podWatch, &api.Pod{}, cache.NewUndeltaStore(source.send, cache.MetaNamespaceKeyFunc), 0).RunUntil(sourceFinished)
-	go source.mergeAndForward()
+	// reflect changes from the watch into a chan, filtered to include only mirror pods
+	// (have an ConfigMirrorAnnotationKey attr)
+	cache.NewReflector(
+		podWatch,
+		&api.Pod{},
+		cache.NewUndeltaStore(source.send, cache.MetaNamespaceKeyFunc),
+		0,
+	).RunUntil(stop)
 }
 
+// send is an update callback invoked by NewUndeltaStore
 func (source *sourceMesos) send(objs []interface{}) {
-	var mirrors []*api.Pod
+	var (
+		pods     = make([]*api.Pod, 0, len(objs))
+		podNames = make(map[podName]string, len(objs))
+	)
+
 	for _, o := range objs {
 		p := o.(*api.Pod)
+		addPod := false
 		if _, ok := p.Annotations[kubetypes.ConfigMirrorAnnotationKey]; ok {
-			mirrors = append(mirrors, p)
+			// pass through all mirror pods
+			addPod = true
+		} else if rpod, err := source.registry.Update(p); err == nil {
+			// pod is bound to a task, and the update is compatible
+			// so we'll allow it through
+			addPod = true
+			podNames[podName{p.Namespace, p.Name}] = rpod.Task()
+		} else {
```

Comment:
probably worth adding an `else if rpod != nil { log something special about an update that failed }`